client/matching/loadbalancer.go

```diff
@@ -135,6 +140,130 @@ func (lb *defaultLoadBalancer) pickPartition(
 		return taskQueue.GetName()
 	}
 
-	n := max(1, nPartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
+	n := max(1, lb.nWritePartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
 	return tqName.WithPartition(rand.Intn(n)).FullName()
 }
+
+// PickReadPartition picks a partition for poller to poll task from, and keeps load balanced between partitions.
+// Caller is responsible to call pollToken.Release() after complete the poll.
+func (lb *defaultLoadBalancer) PickReadPartition(
+	namespaceID namespace.ID,
+	taskQueue taskqueuepb.TaskQueue,
+	taskQueueType enumspb.TaskQueueType,
+	forwardedFrom string,
+) *pollToken {
+	if forwardedFrom != "" || taskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// no partition for sticky task queue and forwarded request
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	parsedName, err := tqname.Parse(taskQueue.GetName())
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	key := taskQueueKey{NamespaceID: namespaceID, Name: parsedName, Type: taskQueueType}
+
+	lb.lock.Lock()
+	tqlb, ok := lb.taskQueueLBs[key]
+	if !ok {
+		tqlb = newTaskQueueLoadBalancer(key)
+		lb.taskQueueLBs[key] = tqlb
+	}
+	lb.lock.Unlock()
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
+
+	namespaceName, err := lb.namespaceIDToName(namespaceID)
+	if err == nil {
+		partitionCount = lb.nReadPartitions(string(namespaceName), parsedName.BaseNameString(), taskQueueType)
+	}
+
+	return tqlb.pickReadPartition(partitionCount, lb.forceReadPartition())
+}
+
+func newTaskQueueLoadBalancer(key taskQueueKey) *tqLoadBalancer {
+	return &tqLoadBalancer{
+		taskQueue: key,
+		lock:      sync.Mutex{},
+	}
+}
+
+func (b *tqLoadBalancer) pickReadPartition(partitionCount int, forcedPartition int) *pollToken {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// ensure we reflect dynamic config change if it ever happens
+	b.ensurePartitionCountLocked(partitionCount)
+
+	partitionID := forcedPartition
+
+	if partitionID < 0 {
+		partitionID = b.pickReadPartitionWithFewestPolls(partitionCount)
+	}
+
+	b.pollerCounts[partitionID]++
+
+	return &pollToken{
+		fullName:    b.taskQueue.Name.WithPartition(partitionID).FullName(),
+		partitionID: partitionID,
+		balancer:    b,
+	}
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) pickReadPartitionWithFewestPolls(partitionCount int) int {
+	// pick a random partition to start with
+	startPartitionID := rand.Intn(partitionCount)
+	pickedPartitionID := startPartitionID
+	minPollerCount := b.pollerCounts[pickedPartitionID]
+	for i := 1; i < partitionCount && minPollerCount > 0; i++ {
```

Comment:
Is there any reason not to use a min-heap here? We also look to have a priority queue impl in our common package

Reply:
The only reason would be simplicity I think. The min-heap implementation should have the following criteria to be a replacement:
- Ability to increase the priority of any given entry (needed for release)
- Ideally, randomized: when multiple minimums exist each should have the same probability of being popped 

Honestly, I'd not expect any min-heap implementation beat the runtime of a simple for loop for the small cardinality that we are looking at.